Lib/ufo2ft/outlineCompiler.py

```diff
@@ -873,6 +880,46 @@ def setupTable_post(self):
         post.minMemType1 = 0
         post.maxMemType1 = 0
 
+    def setupTable_COLR(self):
+        """
+        Compile the COLR table.
+
+        **This should not be called externally.**
+        """
+        if "COLR" not in self.tables:
+            return
+
+        from fontTools.ttLib.tables.C_O_L_R_ import LayerRecord
+        layerInfo = self.ufo.lib[COLOR_LAYERS_KEY]
+        colorLayerLists = {}
+        for baseGlyphName, layers in layerInfo.items():
+            colorLayerLists[baseGlyphName] = [
+                    LayerRecord(layerGlyphName, colorID)
+                        for layerGlyphName, colorID in layers]
+        self.otf["COLR"] = colr = newTable("COLR")
+        colr.version = 0
+        colr.ColorLayers = colorLayerLists
+
+    def setupTable_CPAL(self):
+        """
+        Compile the CPAL table.
+
+        **This should not be called externally.**
+        """
+        if "CPAL" not in self.tables:
+            return
+
+        from fontTools.ttLib.tables.C_P_A_L_ import Color
+        palettes = self.ufo.lib[COLOR_PALETTES_KEY]
```

Comment:
@justvanrossum maybe instead of a list of lists, COLOR_PALETTE_KEY could hold a list of dicts. This way we can support CPAL v1 with optional flags and labels. Each palette can be represented by a dict with a required key `colors` (which is the list of RGBA color tuples), and then three optional keys: `flags` (list of ints), `label` and `entryLabels` (lists of (localised) str).

See `buildCPAL` in https://github.com/fonttools/fonttools/pull/1826

Reply:
Sure, if that makes forward compatibility easier.

Reply:
hm, actually, while it may make make sense to have a dict for each palette (each with an array of `colors` and a `label`), the palette _entry_ labels themselves do not belong to the single palette; they are transversal and thus belong to _all_ of them. So they probably either need their own separate lib key, containing an array of (localized) strings.

(I'd be curious to know if there're any apps that have a UI for displaying to the user such CPALv1 palette labels and palette entry labels...)